File: Python/Largest Rectangle in Histogram.py
class Solution:
    # @param height, a list of integer
    # @return an integer
    def largestRectangleArea(self, height):
		stack = []
		width = []
		top = 0
		area = 0
		newheight = 0
		stack.append(0)
		width.append(0)
		for i in range(0,len(height)+1):
			if  i < len(height):
				newheight = height[i]
			else:
				newheight = -1
			if newheight >= stack[len(stack)-1]:
				stack.append(newheight)
				width.append(1)
			else:
				minV = 2147483647
				wid = 0
				while len(stack) > 0 and stack[len(stack)-1] > newheight:
					minV = min(minV,stack.pop())
					wid = wid + width.pop()
					area = max(area,minV*(wid))
				stack.append(newheight)
				width.append(wid+1)
		return area
		
class Solution:
    # @param height, a list of integer
    # @return an integer
    def largestRectangleArea(self, height):
		stack = []
		area = 0
		newheight = 0
		height.append(0)
		i=0
		while i < len(height):
			if len(stack)==0 or height[i] > height[stack[len(stack)-1]]:
				stack.append(i)
				i = i + 1
			else:
				Min = stack.pop()
				if len(stack) == 0:
					wid = i
				else:
					wid = i - stack[len(stack)-1] -1
				area = max(area,height[Min]*wid)
		return area

File: Java/Largest Rectangle in Histogram.java
class Largest Rectangle in Histogram{
	//similar to brute force, for every eligible col calc the maxAre form with left side col
	//if right col higher then curr col the acra must be larger that form by curr col
	public int largestRectangleArea(int[] height) {
		int maxV = 0;
        for(int i=0;i<height.length;i++){
        	if(i<height.length-1 && height[i] <= height[i+1])
                continue;
            int minH = height[i];
            for(int j=i;j>=0;j--){
            	minH = Math.min(minH,height[i]);
            	maxV = Math.max(maxV,minH*(i-j+1));
            }
        }
        return maxV;
    }
    //use a stack to maintain the index of height, when curr col's height less then
    public int largestRectangleArea(int[] height) {
    	int maxV = 0;
    	LinkedList<Integer> stack = new LinkedList<Integer>();
    	int i=0;
    	int[] h = new int[height.length + 1];
        h = Arrays.copyOf(height, height.length + 1);
    	while(i < h.length){
    		if(stack.isEmpty() || h[i] >= h[stack.peek()]){
    			stack.push(i);
    			i++;
    		}
    		else{
    			int minH = h[stack.pop()];
    			if(stack.isEmpty()){
    				maxV = Math.max(maxV,minH*(i));
    			}
    			else{
    				maxV = Math.max(maxV,minH*(i-stack.peek()-1));
    			}

    		}
    	}
    	return maxV;
    }
}